core/client/fs/src/main/java/alluxio/client/block/stream/BufferCachingGrpcDataReader.java

```diff
@@ -0,0 +1,249 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.block.stream;
+
+import alluxio.client.file.FileSystemContext;
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.conf.PropertyKey;
+import alluxio.grpc.ReadRequest;
+import alluxio.grpc.ReadResponse;
+import alluxio.network.protocol.databuffer.DataBuffer;
+import alluxio.network.protocol.databuffer.NioDataBuffer;
+import alluxio.resource.CloseableResource;
+import alluxio.wire.WorkerNetAddress;
+import alluxio.resource.LockResource;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A gRPC data reader that responsible for reading from a specific block.
+ *
+ * The current implementation cached the block data from the beginning to
```

Comment:
This makes is sound like this is just like the normal grpc data reader, but just caches the data.

If that is the case, why can't we just add caching option to the grpc data reader? Or, wrap the existing grpc data reader? This looks like a reimplementation, with caching functionality.

Also, could you briefly describe the relationship between this `BufferCachingGrpcDataReader` and `SharedGrpcDataReader `?

Reply:
what you described is correct. We choose to reimplement to ensure no existing code path is modified.

`SharedGrpcDataReader` is re-using existing DataReaders to reduce the cost on creation and close. 
`BufferCachingGrpcDataReader` is providing the caching function.